app/models/access_token.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+class AccessToken < ApplicationRecord
+  belongs_to :user
+
+  # Bearerトークンとしての利用を想定しているため、フォーマットはBearerトークンのシンタックスを満たす必要がある。
+  # https://datatracker.ietf.org/doc/html/rfc6750#section-2.1
+  #
+  # 上記では、トークンの文字列長は1以上であればよいことになっているが、短すぎると総当たりで突破されないとも
+  # 限らないので 保険として、 `=` によるパディング以外の部分に関して、ある程度の長さを要求する。
+  validates :token, format: { with: %r{\A[\w\-.~+/]{16,}=*\Z} }, uniqueness: true
+  validates :scope, format: { with: /\A(READ|WRITE)( (READ|WRITE)*)\Z/ }
+  validates :expires_in, numericality: { greater_than_or_equal_to: 0 }
+
+  # 新規レコードの場合かつ属性が引数で明示的に設定されていない場合はデフォルト値を設定する。
+  def initialize(attributes = nil)
+    super attributes
+
+    return unless new_record? && attributes.is_a?(Hash)
```

Comment:
ここにくるときって必ず new_record じゃないっけ

Reply:
メソッドの説明では、必ずしもtrueではないように読み取れますが、#initialize内ではそうでもないんですかね。

https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-new_record-3F

Reply:
>Returns true if this object hasn't been saved yet – that is, a record for the object doesn't exist in the database yet; otherwise, returns false.

`AccessToken.new` して、オブジェクトが生成している最中にデータベースに保存されていることってあるのかな。
どんな場合に new_record じゃなくなるという想定をしていました?

Reply:
DBに保存されているレコードを読み出してインスタンス化することを考えていました。でも、親メソッドを見てみたら、がっつりまだ保存されていないときに呼ばれると書いてありますね。実際、コードを見てみてもそのように見受けられました。

> New objects can be instantiated as either empty (pass no construction parameter) or pre-set with attributes but not yet saved (pass a hash with key names matching the associated table column names).

https://api.rubyonrails.org/classes/ActiveRecord/Core.html#method-c-new

というわけで僕の知識不足でした。失礼しました。

Reply:
脱線: DB に保存されているレコードを読み出す場合、どうやってるんだろうねえ。僕もあまり詳しくないので、今度時間が取れたら調べてみます。
おそらく new した後に (`#initialize` が動いたあとに) 読み込んだ属性や `@new_record` フラグを上書きしているんだと思いますが…

Reply:
知識不足のほかに勘違いもありました。

#after_initializeは、.findの場合でも.newの場合でも、両方で呼ばれるとありました。

> Lastly an after_find and after_initialize callback is triggered for each object that is found and instantiated by a finder, with after_initialize being triggered after new objects are instantiated as well.

https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html

なので、これを利用してデフォルト値を設定する場合は、#new_record?による判定を入れたほうがいいという話があったのですが、そちらと混同してしまいました。

https://qiita.com/supercell1023koeda/items/46aad5929491acd20108

Reply:
> 脱線: DB に保存されているレコードを読み出す場合、どうやってるんだろうねえ。僕もあまり詳しくないので、今度時間が取れたら調べてみます。
> おそらく new した後に (`#initialize` が動いたあとに) 読み込んだ属性や `@new_record` フラグを上書きしているんだと思いますが…

こちらにはnewではなくallocateを使うとありますね。そして、まさに前述の僕の勘違いを正してくれる情報ですね。#after_initializeを調べていたときに見たはずなのですが。目の前に答えがあったのに見逃してました...。

https://blog.dalethatcher.com/2008/03/rails-dont-override-initialize-on.html